### ISSUE TITLE
SignIn Ineffective Custom Button Implementation Pattern

### ISSUE THREAD

**Issue by johnhbenetech** [benetech/MathShare#790]
There are several accessibility problems exhibited by the buttons to initiate the OAuth flow for either Google or Microsoft: 
 
• The Google control is not marked up as any type of clickable control at all, meaning that its purpose and actionable nature is not accessibly communicated. 
• The Google control has two child controls — a < img> with an alt attribute and a < span> with the text “Google” — both of which act as click targets.  This leaves screen reader users confused as to which they should activate, or why there are two controls representing the same action. 
• The alt text of the Google and Microsoft logos includes the word “logo”, which is inappropriate in the context of choosing a third-party OAuth provider from a list.
 • The names of the two third party services are entirely in lower case in the alt text. 
• The Google control includes the name of the service on-screen, but the Microsoft control only includes the logo. 
• The Microsoft control has a role of “button”, even though it causes browser navigation to occur and therefore should be a link. 
• Both controls cause a new browser tab or window to open when activated, but this is not communicated to anybody. 

Suggested Solutions
Apply the Correct Role 

Explicitly mark up each control as a link.  Ideally this should be achieved by using native < a> elements (see: Incorrect Button and Link Keyboard Behavior).  Otherwise, a role attribute with a value of “link” should be added to the controls’ < div> containers (and the existing role of “button” removed from the Microsoft control). 

Provide Accessible Labels 
 
Determine a consistent approach of providing a meaningful, accessible label for each control.  Whether the label is displayed on-screen or not represents a design decision, but that decision should apply to both controls.  We recommend that the label should appear on-screen to allow users of speech input technology to easily activate their login control of choice. 
 
We would suggest the following implementation pattern: 
 
• Mark the existing < img> elements as decorative, by setting their alt attributes to an empty string ("").  Note that the string should contain no whitespace or other characters. 
• Ensure that both controls include a textual element (such as a < span> or simple child text node) which contains the name of the OAuth provider using correct title case. 
 
Indicate New Tab/Window 
 
Append text to the label of each link which indicates that activating the control will cause a new browser tab or window to be opened.  For example, “ (opens in new window)” or “ (opens in new tab)”. 
 
Note the space before the opening parenthesis, which is necessary to avoid a screen reader speaking the current and additional text as a single unit.  This space should be implemented using a non-breaking space HTML entity (“& nbsp;”) to maximize compatibility across browser and screen reader combinations.

**Comment by johnhbenetech** [comment]
Proposed Code Sample 1: Native <a> Elements (Recommended) 
 
File: Src/components/SignIn/index.js 
 
Note: given that the aim is consistency across both controls, it may be appropriate to refactor the two methods in the code sample into a single one with ID, service name, logo URL, etc. as parameters.

Google Control 
 
`renderGoogleBtn = () => (  <a id={this.GOOGLE_SIGN_IN} href="#" style={{ … }} className="…">   <div className="abcRioButtonContentWrapper">    <div className="abcRioButtonIcon" style={{ padding: 10 }}>     <div style={{ … }} className="abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18">      <img src={googleLogo} alt="" />     </div>    </div>    <span style={{ … }} className="abcRioButtonContents">     <span id="…">Google</span><span>&nbsp;(opens in new window)</span>    </span>   </div>  </a> ); `

**Comment by johnhbenetech** [comment]
Microsoft Control 
 
`renderMicrosoftBtn = () => (  <a id={this.MS_SIGN_IN} href="#" className={`…`} onClick={this.startMicrosoftSignIn}>   <div className="abcRioButtonContentWrapper">    <div className="abcRioButtonIcon" style={{ padding: 10 }}>     <div style={{ … }} className="abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18">      <img src={microsoftLogo} alt="" />     </div>    </div>    <span style={{ … }} className="abcRioButtonContents">     <span id="…">Microsoft</span><span>&nbsp;(opens in new window)</span>    </span>   </div>  </a> ) `

**Comment by johnhbenetech** [comment]
Proposed Code Sample 2: Custom Link Controls (Not Recommended) 
 
File: Src/components/SignIn/index.js 

Google Control 
 
`renderGoogleBtn = () => (  <div id={this.GOOGLE_SIGN_IN} role="link" tabIndex={0} style={{ … }} className="…">   <div className="abcRioButtonContentWrapper">    <div className="abcRioButtonIcon" style={{ padding: 10 }}>     <div style={{ … }} className="abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18">      <img src={googleLogo} alt="" />     </div>    </div>    <span style={{ … }} className="abcRioButtonContents">     <span id="…">Google</span><span>&nbsp;(opens in new window)</span>    </span>   </div>  </div> );`

Microsoft Control 
 
`renderMicrosoftBtn = () => (  <div id={this.MS_SIGN_IN} role="link" tabIndex={0} className={`…`} onClick={this.startMicrosoftSignIn}>   <div className="abcRioButtonContentWrapper">    <div className="abcRioButtonIcon" style={{ padding: 10 }}>     <div style={{ … }} className="abcRioButtonSvgImageWithFallback abcRioButtonIconImage abcRioButtonIconImage18">      <img src={microsoftLogo} alt="" />     </div>    </div>    <span style={{ … }} className="abcRioButtonContents">     <span id="…">Microsoft</span><span>&nbsp;(opens in new window)</span>    </span>   </div>  </div> ) `

**Comment by rupeshparab** [comment]
I am planning to do this after deploying the new OAuth backend

**Comment by benetechMatt** [comment]
button look and action works as expected for signing in. When I tested it, I was given the message that I successfully signed in but was not taken to the dashboard as expected.  I do like the success and failure messages that display in the upper right corner.

not sure if this is just a temporary issue with signing

**Comment by johnhbenetech** [comment]
@rupeshparab Seems like if you sign in on homepage it should automatically take you to the dashboard.

@jscholes Is there anything special for this case of a page transition? It's not directly driven by a button or link click in our platform. You get sent away to the Oauth flow and then when you come back you'll be in a different page then you left (we will still announce the new page as in #742).

**Comment by jscholes** [comment]
@johnhbenetech - the user will be completing the OAuth flow elsewhere (in a new window or tab?).  So when they're done, their browser will either:

* navigate away from the OAuth provider and then back into MathShare; or
* close the OAuth window and move focus back to the MathShare tab.

In both of those instances, a screen reader will announce something.  The correctness of that announcement is yet to be determined.  For instance, when the OAuth window is dismissed, the page title may still say "Sign In".  So at that point, handling an accessible transition to the dashboard would be important.

Likewise, if a user has previously authenticated with a given provider, the OAuth flow may fly by without requiring any input from them.  That also makes it a good idea to always use accessible transitions within MathShare itself.

**Comment by rupeshparab** [comment]
@johnhbenetech 
Currently the backend redirects to the page from which Sign In was opened.
Will this redirection from HomePage to Dashboard happen only on Sign In completion or should the user be restricted from accessing the landing page if he is authenticated.

@jscholes 
As this transition (Redirection from Oauth provider to MathShare) is page load it will be announced by the browser.
Also, we are focusing a success/failure alert after redirection.

**Comment by jscholes** [comment]
@rupeshparab
> As this transition (Redirection from Oauth provider to MathShare) is page load it will be announced by the browser.

Correct, and I'm not worried about that.  I am concerned that if the Sign In page is still showing when that happens, even for a second or so, the screen reader will indicate that page as the new focus point.  At that point, an explicitly announced and handled transition becomes necessary on MathShare's side.

> Also, we are focusing a success/failure alert after redirection.

Are you moving focus to the heading, and does it have `tabindex="-1"`?

**Comment by johnhbenetech** [comment]
@benetechMatt Please see @rupeshparab question above. Basically if the user has previously logged in, leaves mathshare completely, and then returns (assuming their auth token hasnt expired) - what should happen if they try to visit mathshare.benetech.org?

1. They see the homepage, with the Go to App link where sign is was
2. They never see the homepage and are immediately sent to dashboard. A simple implementation of this would mean a logged in user could never see the homepage.

**Comment by rupeshparab** [comment]
@jscholes 
Currently on QA (https://mathshare-qa.diagramcenter.org/) the the user gets redirected directly to the page he was before clicking on Sign In
I noticed that if a user reloads the page while still on Sign In or directly tries to open 'Sign In' from the address bar, the redirect url sent to the backend is same as the 'Sign In' page.
I have fixed it so that even in these 2 cases user won't be redirected to 'Sign In' and will be redirected by default to '/app'

> Are you moving focus to the heading, and does it have tabindex="-1"?
Yes, I am focusing on the heading of the alert and I have given it a tabIndex

**Comment by jscholes** [comment]
@rupeshparab - all working here, even when the OAuth flow has already been completed and hence just takes place within the same browser window.

**Comment by rupeshparab** [comment]
Sure, will clear the message such that there is only 1 login/logout related message at a time

**Comment by benetechMatt** [comment]
@rupeshparab If the user is still signed in and they go to mathshare.benetech.org they should be taken to the /app instead of the homepage/landing page.

**Comment by jscholes** [comment]
@rupeshparab - there is still some inconsistency in that the Google logo displays a textual label on-screen, but the Microsoft one does not.

@benetechMatt - if there are outstanding issues with sign in behaviour which don't relate to the implementation of the OAuth provider links, please create separate tickets.

**Comment by rupeshparab** [comment]
@jscholes 
Currently, Google logo is an image and there is text 'Google' to its right. Microsoft button has just an image which contains the logo and the text in order to match its typeface. I have also included the 'Microsoft' text in a screen reader only span.
Do you want me to change the Microsoft image so that its container has separate logo and text?

**Comment by jscholes** [comment]
@rupeshparab - yes please.  Images of text should be avoided in most instances.  They can cause all sorts of issues for low vision users, and speech recognition users will try to speak the text only to possibly find that it doesn't work.

**Comment by rupeshparab** [comment]
I have changed the Microsoft button to include text, changed the alerts for login such that there is just 1 login related alert at a time and also made sure that if a user navigates to the signin page all signin alerts get cleared.

**Comment by jscholes** [comment]
Fixed in https://github.com/benetech/MathShare/commit/320b55d8ea4508354cb64779351eebbb2414fb76.
